### PR TITLE
PIM-7090: add completeness support on export builder

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

This PR adds new filters on the product query builder to be able to filter on all_complete product models and all_incomplete product models. 

We changed the way we index the the completeness level of children as previously it was a nightmare to understand with double negation. We took the chance to change that as we had to re-index anyway.
Nothing changes for PQB users, this is only internal to ES index.

We also fixed the AT_LEAST_COMPLETE and AT_LEAST_INCOMPLETE operator that where not consistent with the previous operators (we never use underscore in operators). We spend a lot of time to understand what was going on and it was because of this inconsistency. So maybe integrators we feel the same. Same as before, this is "internal" as integrators should use the constant. It could have an impact on the API but as any changes on the PQB does. IMHO, this is ok.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Yes
| Added legacy Behats               | No
| Added acceptance tests            | No
| Added integration tests           | Yes
| Changelog updated                 | Yes
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Yes
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Component/Catalog/spec/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizerSpec.php [hunk 1]
```diff
@@ -146,6 +181,7 @@ function it_normalizes_a_root_product_model_fields_and_values(
 
         $productModel->getCategoryCodes()->willReturn(['category_A', 'category_B']);
 
+
```
Comment (christophe-chausseray) [inline src/Pim/Component/Catalog/spec/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizerSpec.php]: Useless blank line

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+use Pim\Component\Catalog\Query\Filter\Operators;
+
+/**
+ * Test that the right product and product model are found depending
+ *  - the completeness for a variant product
+ *  - the number of complete variant product
+ *
+ * Caution: We use product and product model query builder here
+ *
+ * @author    Arnaud Langlade <arnaud.langlade@akeneo.com>
```
Comment (christophe-chausseray) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php]: The author is you, no? :)

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+class ExportCompletenessFilterIntegration extends AbstractProductQueryBuilderTestCase
+{
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->getFromTestContainer('akeneo_integration_tests.catalog.fixture.completeness_filter')
+            ->loadProductModelTree();
+
+        sleep(2);
```
Comment (christophe-chausseray) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php]: Is it mandatory this sleep after the setup?

Comment (juliensnz) [reply src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php]: No idea, it's a copy and paste

File: CHANGELOG-2.2.md [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
+
 ## BC breaks
 
 ### Interfaces
 
 - PIM-7163: Add `Pim\Bundle\UserBundle\Entity\UserInterface::setPhone` and `Pim\Bundle\UserBundle\Entity\UserInterface::getPhone`
 
+## Migration
+
+- PIM-7090: You need to re-index product and product model indexes in elastic search
```
Comment (SamirBoulil) [inline CHANGELOG-2.2.md]: You could do a message just a little more informative like in 2.0: 
```- New data has been indexed in Elasticsearch. Please re-index the products and product models by launching the commands `pim:product:index --all -e prod` and `pim:product-model:index --all -e prod`.
```

Anyway we'll think next week of a way to run the index commands during the migration process ?

Comment (juliensnz) [reply CHANGELOG-2.2.md]: yeah, good idea
will do :)

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Test the AT LEAST COMPLETE filter on all locale
+     * @group do
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Field/ExportCompletenessFilterIntegration.php]: remove the @ groups

File: src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml [hunk 6]
```diff
@@ -14,7 +14,6 @@ services:
             - ['ALL']
         tags:
             - { name: 'pim_catalog.elasticsearch.query.product_filter', priority: 30 }
-            - { name: 'pim_catalog.elasticsearch.query.product_model_filter', priority: 30 }
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml]: why ?

File: src/Pim/Component/Catalog/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizer.php [hunk 7]
```diff
@@ -25,8 +26,18 @@ class ProductModelPropertiesNormalizer implements NormalizerInterface, Serialize
...
+    private const FIELD_ALL_INCOMPLETE = 'all_incomplete';
+    private const FIELD_ALL_COMPLETE = 'all_complete';
     private const FIELD_ANCESTORS = 'ancestors';
 
+    /** @var CompleteFilterInterface */
+    private $completenessGridFilterQuery;
+
+    public function __construct(CompleteFilterInterface $completenessGridFilterQuery)
+    {
+        $this->completenessGridFilterQuery = $completenessGridFilterQuery;
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizer.php]: It's really wierd to have a normalizer for the ES index depending on something called "Grid" and "Filter" at the same time. It seems fishy.

Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizer.php]: Ok, just rename your private to something like completenessFilterQuery (but I thing the naming for the interface is wrong) a filter in the PIM cannot be a doctrine query function, it's too ambiguous with PQB's filters (but that's not to be done in your PR ;) )

Comment (juliensnz) [reply src/Pim/Component/Catalog/Normalizer/Indexing/ProductModel/ProductModelPropertiesNormalizer.php]: Yeah, definitely, the Interface is really badly named. I will change the property name :)

File: src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml (hunk 6 above)
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml]: The final diff shows that it was added :)

**Pull request merged**